problem_builder/instructor_tool.py

```diff
@@ -0,0 +1,359 @@
+        """
+        from .tasks import export_data as export_data_task  # Import here since this is edX LMS specific
+        if self.active_export_task_id:
+            async_result = export_data_task.AsyncResult(self.active_export_task_id)
+            if async_result.ready():
+                self._save_result(async_result)
+
+    def _save_result(self, task_result):
+        """ Given an AsyncResult or EagerResult, save it. """
+        self.active_export_task_id = ''
+        if task_result.successful():
+            if isinstance(task_result.result, dict) and not task_result.result.get('error'):
+                self.display_data = task_result.result['display_data']
+                del task_result.result['display_data']
+                self.last_export_result = task_result.result
+            else:
+                self.last_export_result = {'error': u'Unexpected result: {}'.format(repr(task_result.result))}
+                self.display_data = None
+        else:
+            self.last_export_result = {'error': unicode(task_result.result)}
+            self.display_data = None
+
+    @XBlock.json_handler
+    def get_result_page(self, data, suffix=''):
+        """ Return requested page of `last_export_result`. """
+        paginator = Paginator(self.display_data, PAGE_SIZE)
+        page = data.get('page', None)
+        return {
+            'display_data': paginator.page(page).object_list,
+            'num_results': len(self.display_data),
+            'page_size': PAGE_SIZE
+        }
+
+    def student_view(self, context=None):
+        """ Normal View """
+        if not self.user_is_staff():
+            return Fragment(u'<p>This interface can only be used by course staff.</p>')
+        block_choices = {
+            _('Multiple Choice Question'): 'MCQBlock',
+            _('Rating Question'): 'RatingBlock',
+            _('Long Answer'): 'AnswerBlock',
+        }
+
+        flat_block_tree = self._build_course_tree()
+
+        html = loader.render_template(
+            'templates/html/instructor_tool.html',
+            {'block_choices': block_choices, 'block_tree': flat_block_tree}
+        )
+        fragment = Fragment(html)
+        fragment.add_css_url(self.runtime.local_resource_url(self, 'public/css/instructor_tool.css'))
+        fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/instructor_tool.js'))
+        fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/vendor/underscore-min.js'))
+        fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/vendor/backbone-min.js'))
+        fragment.add_javascript_url(self.runtime.local_resource_url(self, 'public/js/vendor/backbone.paginator.min.js'))
+        fragment.initialize_js('InstructorToolBlock')
+        return fragment
+
+    def _build_course_tree(self):
+        """
+        Return flat tree of blocks belonging to this block's parent course.
+        """
```

Comment:
Awesome, thank you so much!

BTW, we just had a discussion yesterday where it came up that we're planning to add the course blocks API as an XBlock runtime service to help power pluggable navigation efforts:

https://openedx.atlassian.net/wiki/display/TNL/UIBlocks+for+Course+Outline

That being said, I think your approach with hitting Course Blocks API from the client-side is the right one in this case, since it won't adversely affect the XBlock server side rendering for large sequences. I just didn't want you folks to be surprised when the course blocks runtime service pops up.

Take care!